### PR TITLE
feat: configuration generator

### LINKED ISSUE #2004: feat: Generator Configuration
This feature essentially allows developers to create GraphQL configurations automatically. This is done by providing a few inputs to the "ConfigGenerator" and then a new Configuration is automatically generated.

**Technical Requirements**

1. Here is sample CLI integration of this feature:

	```bash
	tailcall generate ./generator.json
	```

2. Proposed Configuration Input

	```rust
	
	enum Operation {
	    Query,
	    Mutation
	}
	
	struct Input {
	    source: Source,
	    field_name: String,
	    operation: Option<Operation>,    
	}
	
	enum Source {
	    URL {
	        url: String, 
	        headers: Option<BTreeMap<String, String>>,        
	        method: Option<Method>, 
	        body: Option<serde_json::Value>
	    },
	    Proto {path: String},
	    Config {url: String}
	}
	
	enum Transformer {
	    TypeMerger {threshold: Option<f32>},
	    AmbiguousType {
	        input: Option<Name>,
	        output: Option<Name>
	    },
	    ConsolidateBaseURL {threshold: Option<f32>},
	    BetterTypeName,
	    TreeShake
	}
	
	struct Name {
	    prefix: Option<String>,
	    postfix: Option<String>
	}
	
	struct Config {
	    inputs: Vec<Input>,
	    transformers: Vec<Transformer>,
	    output: Output,
	    schema: {
	        query: Option<String>,
	        mutation: Option<String>
	    } 
	}
	
	struct Output {
	    path: String,
	    format: Option<Format>
	}
	
	enum Format {
	    GraphQL,
	    JSON,
	    YAML
	}
	```

### PR DESCRIPTION
**Pull request by meskill**
fixes #2004

### CODE REVIEW

**Comment by meskill** [comment]
TODO:
- [ ] resolve paths inside the config and for generated output relative to the config itself
- [ ] generate JSON schema for config generation
- [ ] add snapshot tests for generation
- [ ] ask for confirmation to overwrite if the output file already exists
- [ ] when running generator if the input file doesn't exist prompt to create it and fill it with simple configuration and `$schema` property set to generated JSON schema
- [ ] optional: refactor `Source`, `ImportSource` to remove duplications. Maybe generate something automatically with the help of `strum_macros`

File: src/core/generator/from_json.rs [hunk 1]
```diff
@@ -4,49 +4,126 @@ use url::Url;
...
+        Self {
+            request_samples,
+            type_name_generator,
+            field_name_generator,
+            operation_name: operation_name.to_string(),
+        }
+    }
+}
+
+impl Generate for FromJsonGenerator<'_> {
```
Comment (tusharmath) [inline src/core/generator/from_json.rs:50]: implement Transform instead of Generate

File: examples/generator/config.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+      "import": {
+        "src": "https://jsonplaceholder.typicode.com/users"
+      }
+    }
+  ],
+  "output": {
+    "format": "graphQL",
+    "file": "./output.graphql"
+  }
+}
```
Comment (tusharmath) [inline examples/generator/config.json:18]: Not required. Lets drop this file.

File: src/core/generator/source.rs [hunk 3]
```diff
@@ -1,25 +1,88 @@
 use thiserror::Error;
+use url::Url;
 
 ///
 /// A list of sources from which a configuration can be created
 #[derive(Clone, Copy, PartialEq, Debug, Default)]
-pub enum Source {
+pub enum ImportSource {
```
Comment (tusharmath) [inline src/core/generator/source.rs:7]: Can be moved to cli.

**Review by tusharmath — APPROVED** [review]

**Pull request merged**